diff -ur http-old/include/functions.php http/include/functions.php
--- http-old/include/functions.php	2016-01-22 18:51:34.000000000 +0000
+++ http/include/functions.php	2016-03-01 06:30:24.872047487 +0000
@@ -1,6 +1,4 @@
 <?php
-header("X-XSS-Protection: 0"); // ;-)
-
 // Reload the page
 function reload_page()
 {
@@ -37,8 +35,9 @@
 // Create a session token
 function create_token($username)
 {
-    setcookie("username", $username);
-    setcookie("session", md5($username));
+    $_SESSION["username"] = $username;
+    $_SESSION["session"] = md5($username);
+    $_SESSION["csrf_key"] = openssl_random_psuedo_bytes(16);
 }
 
 // Check a session token
@@ -50,8 +49,7 @@
 // Destroy the session token
 function logout()
 {
-    setcookie("username", "", time()-3600);
-    setcookie("session", "", time()-3600);
+    session_unset(); 
 }
 
 function login($username, $password)
@@ -82,9 +80,9 @@
 // Check if there is a user signed in
 function check_signed_in()
 {
-    if (isset($_COOKIE['username'], $_COOKIE['session']))
+    if (isset($_SESSION['username'], $_SESSION['session']))
     {
-        if (check_token($_COOKIE['username'], $_COOKIE['session']))
+        if (check_token($_SESSION['username'], $_SESSION['session']))
      	{   return True; }
         else
         {
@@ -112,7 +110,7 @@
         {
             print "<blockquote>";
             print "<p>";
-     	    print "<img src='".$row['link']."' />";
+     	    print "<img src='".htmlspecialchars($row['link'])."' />";
             print "</p>";
             print "<footer>";
             print "<cite>";
@@ -123,8 +121,9 @@
             print "<b>Vote count: ".$row['vote']."</b>";
             print "</cite>";
             print "</footer>";
-	    print "<form action='vote.php' method='GET'>";
+            print "<form action='vote.php' method='GET'>";
             print "<input type='hidden' name='vote' value='".$row['user']."' />";
+            print "<input type='hidden' name='key' value='".$_SESSION["csrf_key"]."' />";
             print "<input type='submit' value='Vote for me' />";
             print "</form>";
             print "</blockquote>";
@@ -170,6 +169,8 @@
         {
             $post = $db->prepare("UPDATE image SET link=:link WHERE user=:user");
         }
+        
+        $link = htmlspecialchars($link);
 
         $post->bindParam(':link', $link);
         $post->bindParam(':user', $user);
@@ -246,5 +247,4 @@
         return false;
     }
 }
-
 ?>
diff -ur http-old/index.php http/index.php
--- http-old/index.php	2016-01-22 18:49:07.000000000 +0000
+++ http/index.php	2016-03-01 04:54:07.788691406 +0000
@@ -1,3 +1,6 @@
+<?php
+    session_start();
+?>
 <html lang="en">
   <head>
     <meta charset="utf-8">
@@ -34,7 +37,7 @@
         {
             if (isset($_POST['post']))
             {
-             	$whose = $_COOKIE['username'];
+             	$whose = $_SESSION['username'];
                 $link = $_POST['link'];
 
                 submit_image_link($whose, $link);
diff -ur http-old/vote.php http/vote.php
--- http-old/vote.php	2016-01-22 18:40:18.000000000 +0000
+++ http/vote.php	2016-03-01 06:37:23.875354218 +0000
@@ -1,3 +1,6 @@
+<?php
+    session_start();
+?>
 <html lang="en">
   <head>
     <meta charset="utf-8">
@@ -32,17 +35,17 @@
       <?php
 	if (check_signed_in())
         {
-            if (!isset($_GET['vote']) || check_uniqueness(get_db(),$_GET['vote']))
+            if (!isset($_GET['vote']) || !isset($_GET['key']) || !isset($_SESSION['csrf_key'])|| check_uniqueness(get_db(),$_GET['vote']) || $_SESSION['csrf_key'] != $_GET['key'])
             {
              	print ("<p>Invalid vote!!</p>");
             }
-            elseif ($_COOKIE['username'] == $_GET['vote'])
+            elseif ($_SESSION['username'] == $_GET['vote'])
             {
                 print ("<p>You cannot vote for yourself!!</p>");
             }
             else
             {
-                if(vote($_COOKIE['username'],$_GET['vote'])) 
+                if(vote($_SESSION['username'],$_GET['vote'])) 
                 {
                     header("Location:/");
                 }
